File: CS/GantSolution.Module.Win/Editors/CustomGanttEditor.cs
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Editors;
using DevExpress.ExpressApp.Model;
using DevExpress.XtraGantt;
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using GanttSolution.Module.BusinessObjects;

namespace GanttSolution.Module.Win.Editors {
    [ListEditor(typeof(ITask), false)]
    public class CustomGanttEditor : ListEditor {
        private GanttControl control;
        private object controlDataSource;
        public CustomGanttEditor(IModelListView info) : base(info) { }
        public override SelectionType SelectionType {
            get { return SelectionType.Full; }
        }
        public override IList GetSelectedObjects() {
            if(control == null) {
                return new object[0] { };
            }
            object[] result = new object[1];
            result[0] = control.GetFocusedRow();
            return result;
        }
        private void Control_SelectedIndexChanged(object sender, EventArgs e) {
            OnSelectionChanged();
            OnFocusedObjectChanged();
        }
        private void Control_ItemSelectionChanged(object sender, ListViewItemSelectionChangedEventArgs e) {
            OnSelectionChanged();
        }
        public override void Refresh() {
            if(control == null)
                return;
            try {
                control.BeginUpdate();
                control.DataSource = controlDataSource;
                control.RefreshDataSource();
            }
            finally {
                control.EndUpdate();
            }
        }
        private void DataSource_ListChanged(object sender, ListChangedEventArgs e) {
            Refresh();
        }
        protected override void AssignDataSourceToControl(object dataSource) {
            if(controlDataSource != dataSource) {
                IBindingList oldBindable = controlDataSource as IBindingList;
                if(oldBindable != null) {
                    oldBindable.ListChanged -= new ListChangedEventHandler(DataSource_ListChanged);
                }
                controlDataSource = dataSource;
                IBindingList bindable = controlDataSource as IBindingList;
                if(bindable != null) {
                    bindable.ListChanged += DataSource_ListChanged;
                }
                Refresh();
            }
        }
        protected override object CreateControlsCore() {
            control = new GanttControl();
            foreach(IModelColumn column in Model.Columns) {
                var ganttColumn = new DevExpress.XtraTreeList.Columns.TreeListColumn();
                ganttColumn.Caption = column.Caption;
                ganttColumn.FieldName = column.PropertyName;
                ganttColumn.Name = column.PropertyName + "Column";
                ganttColumn.Visible = true;
                ganttColumn.SortIndex = column.SortIndex;
                control.Columns.Add(ganttColumn);
            }
            control.KeyFieldName = nameof(ITask.Id);
            control.ParentFieldName = nameof(ITask.Parent);
            control.ChartMappings.TextFieldName = nameof(ITask.Name);
            control.ChartMappings.StartDateFieldName = nameof(ITask.StartDate);
            control.ChartMappings.FinishDateFieldName = nameof(ITask.EndDate);
            control.ChartMappings.ProgressFieldName = nameof(ITask.Progress);
            control.ChartMappings.PredecessorsFieldName = nameof(ITask.PredecessorTasks);
            control.AllowTouchGestures = DevExpress.Utils.DefaultBoolean.True;
            control.OptionsBehavior.ScheduleMode = DevExpress.XtraGantt.Options.ScheduleMode.Auto;
            control.OptionsCustomization.AllowModifyTasks = DevExpress.Utils.DefaultBoolean.True;
            control.OptionsCustomization.AllowModifyDependencies = DevExpress.Utils.DefaultBoolean.True;
            control.OptionsCustomization.AllowModifyProgress = DevExpress.Utils.DefaultBoolean.True;
            control.SelectionChanged += Control_SelectedIndexChanged;
            control.FocusedNodeChanged += Control_FocusedNodeChanged;
            control.MouseDoubleClick += Control_MouseDoubleClick;
            control.KeyDown += Control_KeyDown;
            Refresh();
            return control;
        }
        private void Control_FocusedNodeChanged(object sender, DevExpress.XtraTreeList.FocusedNodeChangedEventArgs e) {
            OnSelectionChanged();
            OnFocusedObjectChanged();
        }
        private void Control_MouseDoubleClick(object sender, MouseEventArgs e) {
            if(e.Button == MouseButtons.Left) {
                GanttControlHitInfo hitInfo = control.CalcHitInfo(e.Location);
                if(hitInfo.ChartHitTest?.ItemInfo != null) {
                    OnProcessSelectedItem();
                }
            }
        }
        private void Control_KeyDown(object sender, KeyEventArgs e) {
            if(e.KeyCode == Keys.Enter) {
                OnProcessSelectedItem();
            }
        }
        public override object FocusedObject {
            get {
                return control?.GetFocusedRow();
            }
        }
        public override void Dispose() {
            controlDataSource = null;
            base.Dispose();
        }
    }
}

File: VB/GanttSolution.Module.Win/Editors/CustomGanttEditor.vb
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Editors
Imports DevExpress.ExpressApp.Model
Imports DevExpress.XtraGantt
Imports System.ComponentModel
Imports GanttSolution.Module.BusinessObjects

Namespace GanttSolution.Module.Win.Editors
    <ListEditor(GetType(ITask), False)>
    Public Class CustomGanttEditor
        Inherits ListEditor
        Private ganttControl As GanttControl
        Private controlDataSource As Object
        Public Sub New(ByVal info As IModelListView)
            MyBase.New(info)
        End Sub
        Public Overrides ReadOnly Property SelectionType() As SelectionType
            Get
                Return SelectionType.Full
            End Get
        End Property
        Public Overrides Function GetSelectedObjects() As IList
            If ganttControl Is Nothing Then
                Return New Object() {}
            End If
            Dim result(0) As Object
            result(0) = ganttControl.GetFocusedRow()
            Return result
        End Function
        Private Sub Control_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs)
            OnSelectionChanged()
            OnFocusedObjectChanged()
        End Sub
        Private Sub Control_ItemSelectionChanged(ByVal sender As Object, ByVal e As ListViewItemSelectionChangedEventArgs)
            OnSelectionChanged()
        End Sub
        Public Overrides Sub Refresh()
            If ganttControl Is Nothing Then
                Return
            End If
            Try
                ganttControl.BeginUpdate()
                ganttControl.DataSource = controlDataSource
                ganttControl.RefreshDataSource()
            Finally
                ganttControl.EndUpdate()
            End Try
        End Sub
        Private Sub DataSource_ListChanged(ByVal sender As Object, ByVal e As ListChangedEventArgs)
            Refresh()
        End Sub
        Protected Overrides Sub AssignDataSourceToControl(ByVal dataSource As Object)
            If controlDataSource IsNot dataSource Then
                Dim oldBindable As IBindingList = TryCast(controlDataSource, IBindingList)
                If oldBindable IsNot Nothing Then
                    RemoveHandler oldBindable.ListChanged, AddressOf DataSource_ListChanged
                End If
                controlDataSource = dataSource
                Dim bindable As IBindingList = TryCast(controlDataSource, IBindingList)
                If bindable IsNot Nothing Then
                    AddHandler bindable.ListChanged, AddressOf DataSource_ListChanged
                End If
                Refresh()
            End If
        End Sub
        Protected Overrides Function CreateControlsCore() As Object
            ganttControl = New GanttControl()
            For Each column As IModelColumn In Model.Columns
                Dim ganttColumn = New DevExpress.XtraTreeList.Columns.TreeListColumn()
                ganttColumn.Caption = column.Caption
                ganttColumn.FieldName = column.PropertyName
                ganttColumn.Name = column.PropertyName & "Column"
                ganttColumn.Visible = True
                ganttColumn.SortIndex = column.SortIndex
                ganttControl.Columns.Add(ganttColumn)
            Next column
            ganttControl.KeyFieldName = NameOf(ITask.Id)
            ganttControl.ParentFieldName = NameOf(ITask.Parent)
            ganttControl.ChartMappings.TextFieldName = NameOf(ITask.Name)
            ganttControl.ChartMappings.StartDateFieldName = NameOf(ITask.StartDate)
            ganttControl.ChartMappings.FinishDateFieldName = NameOf(ITask.EndDate)
            ganttControl.ChartMappings.ProgressFieldName = NameOf(ITask.Progress)
            ganttControl.ChartMappings.PredecessorsFieldName = NameOf(ITask.PredecessorTasks)
            ganttControl.AllowTouchGestures = DevExpress.Utils.DefaultBoolean.True
            ganttControl.OptionsBehavior.ScheduleMode = DevExpress.XtraGantt.Options.ScheduleMode.Auto
            ganttControl.OptionsCustomization.AllowModifyTasks = DevExpress.Utils.DefaultBoolean.True
            ganttControl.OptionsCustomization.AllowModifyDependencies = DevExpress.Utils.DefaultBoolean.True
            ganttControl.OptionsCustomization.AllowModifyProgress = DevExpress.Utils.DefaultBoolean.True
            AddHandler ganttControl.SelectionChanged, AddressOf Control_SelectedIndexChanged
            AddHandler ganttControl.FocusedNodeChanged, AddressOf Control_FocusedNodeChanged
            AddHandler ganttControl.MouseDoubleClick, AddressOf Control_MouseDoubleClick
            AddHandler ganttControl.KeyDown, AddressOf Control_KeyDown
            Refresh()
            Return ganttControl
        End Function
        Private Sub Control_FocusedNodeChanged(ByVal sender As Object, ByVal e As DevExpress.XtraTreeList.FocusedNodeChangedEventArgs)
            OnSelectionChanged()
            OnFocusedObjectChanged()
        End Sub
        Private Sub Control_MouseDoubleClick(ByVal sender As Object, ByVal e As MouseEventArgs)
            If e.Button = MouseButtons.Left Then
                Dim hitInfo As GanttControlHitInfo = ganttControl.CalcHitInfo(e.Location)
                If hitInfo.ChartHitTest?.ItemInfo IsNot Nothing Then
                    OnProcessSelectedItem()
                End If
            End If
        End Sub
        Private Sub Control_KeyDown(ByVal sender As Object, ByVal e As KeyEventArgs)
            If e.KeyCode = Keys.Enter Then
                OnProcessSelectedItem()
            End If
        End Sub
        Public Overrides Property FocusedObject() As Object
            Get
                Return ganttControl?.GetFocusedRow()
            End Get
            Set(value As Object)
            End Set
        End Property
        Public Overrides Sub Dispose()
            controlDataSource = Nothing
            MyBase.Dispose()
        End Sub
    End Class
End Namespace

